gtests/net/tcp/common/set_sysctls.py

```diff
@@ -1,24 +1,46 @@
-#!/usr/bin/python
+#!/usr/bin/env python3
+#
+# Copyright 2012 Google Inc. All Rights Reserved.
+
+"""Sets sysctl values and writes a file that restores them.
+
+The arguments are of the form "<proc-file>=<val>" separated by spaces.
+The program first reads the current value of the proc-file and creates
+a shell script named "/tmp/sysctl_restore_${PPID}.sh" which restores the values
+when executed. It then sets the new values
+"""
+
+__author__ = ('brakmo@google.com (Lawrence Brakmo)')
 
-import commands
 import os
+import subprocess
 import sys
 
-pppid = int(os.popen("ps -p %d -oppid=" % os.getppid()).read().strip())
-filename = '/tmp/sysctl_restore_%d.sh' % pppid
+# Sometimes shell skips fork() as an optimization.
+# https://unix.stackexchange.com/questions/466496/
+# If your shell does so, define SHELL_SKIPS_FORK=1
```

Comment:
Interesting! For my education, on which OS distribution and shell did you run into this issue?

Reply:
On Debian 11 and Ubuntu, /bin/sh points to `dash`, which doesn't skip fork(), so I get process tree like:
```
virtme-init─┬─bash───packetdrill─┬─sh───python3
            │                    │ ^^ Here's /bin/sh by system(3)
            │                    └─{packetdrill}
            └─systemd-udevd───2*[systemd-udevd]
```
In this case, `os.getppid()` returns pid of `sh`, not `packetdrill`, so one additional step is needed to find **pppid** (line 26 below). This is the original behavior of this script (the open source version). I believe that's the reason Willem introduced **pppid** in commit 51a1a06b.

However, in Google, `/bin/sh` points to `bash`, which **does** skip fork() for simple command like this. So `os.getppid()` returns pid of `packetdrill`:

```
virtme-init─┬─bash───packetdrill─┬─python3
            │                    └─{packetdrill}
            └─systemd-udevd───2*[systemd-udevd]
```
So the internal version of this script uses `os.getppid()` alone, no need to call `ps` command.

I didn't know any programmtic way to detect shell's behavior, or disable this optimization in `bash`.
I introduced `SHELL_SKIPS_FORK` so that I could test inside Google before uploading to github.com.
This is far from ideal, and I have no idea how to keep this part of packetdrill upstream and internally in sync :-(

Suggestions?

Reply:
Thanks for the detailed explanation! That's very clear and very helpful.

It seems like it will not be a good user experience if packetdrill users have to know whether their shell skips fork, and thus know whether to set the SHELL_SKIPS_FORK environment variable. :-)

I would suggest that we change the set_sysctls.py approach so that it does not rely on PPID. I think Larry Brakmo just used that for the set/restore scheme because it was easy. But now that commonly used shells have diverged in this way, it's no longer easy, so I think we should just change approaches. :-)

Here's a proposal: I'd suggest a small patch to packetdrill to export a  unique ID for the test into an environment variable, and set_sysctls.py and the restore command can both use that to coordinate. We could use many different approaches to generate the testid, but since we know the pid is simple and works, I guess we can start with that. So how about something like: 

1: packetdrill can run something like:
 (a) the following in state_new() in run.c:
  asprintf(&state->test_id, "%d", getpid());
  setenv("PACKETDRILL_TEST_ID", state->test_id, 1);
 (b) the following in state_free():
  free(state->test_id);
 (c) test_id can be defined in the state struct in run.h with something like:
   char *test_id; /* opaque id string for current test */

2: set_sysctls.py can run:
  filename = '/tmp/sysctl_restore_%s.sh' % os.getenv('PACKETDRILL_TEST_ID')

3: for scripts that restore the sysctl values, we can change:
    `/tmp/sysctl_restore_${PPID}.sh`
  to:
    `/tmp/sysctl_restore_${PACKETDRILL_TEST_ID}.sh`

  You should be able to change all of the existing sysctl_restore_ invocations with something like:
` find gtests/net/ -name '*.pkt' |  xargs perl -i.orig -p -e 's|/tmp/sysctl_restore_\${PPID}.sh|/tmp/sysctl_restore_\${PACKETDRILL_TEST_ID}.sh|g'
`
 
I would suggest testing this briefly on one test upstream and then one test in 9xx. If it works on one test upstream and 9xx, then I'd suggest testing it for all //net/tcp/... and //net/gre/... and //net/packetdrill/... tests in 9xx, and then sending it for review in 9xx. Then we can send it upstream. Thanks!

@wdebruij  for visibility and in case Willem has other ideas.

Reply:
Thanks for the suggestion! I updated this pull request:
1. call `setenv(name, value, 1)` in main(). According to `man 3 setenv` makes copies of the strings pointed to by name and value (by contrast with `putenv(3)`), so no need to add a `test_id` field to struct `state`.
2. updated `set_sysctls.py` as you described.
3. no need to update `.pkt` files as long as `PACKETDRILL_TEST_ID` is set to pid of packetdrill. Because the substitution of `${PPID}` happens inside the `sh` started by `system(3)`, its `PPID` points to packetdrill no matter what shell we are using.

I tested this approach works for both `bash` and `dash` on my own machine. I will set it for review in 9xx too.